src/pudl/analysis/record_linkage/eia_ferc1_record_linkage.py

```diff
@@ -35,25 +35,277 @@
+    ferc1_df = inputs.get_plants_ferc1().reset_index()
+    eia_df = inputs.get_plant_parts_eia_true().reset_index()
+    block_col = "plant_id_report_year_util_id"
+    out = ferc1_df.merge(
+        eia_df, how="inner", on=block_col, suffixes=("_ferc1", "_eia")
+    ).set_index(["record_id_ferc1", "record_id_eia"])
+    return out
+
+
+@op
+def get_train_pairs_df(inputs):
+    """Get a dataframe with possible FERC to EIA record pairs from training data.
+
+    Merge the FERC and EIA records on ``block_col`` to generate possible
+    record pairs for the matching model.
+
+    Arguments:
+        inputs: :class:`InputManager` object.
+    """
+    ferc1_df = inputs.get_train_ferc1().reset_index()
+    eia_df = inputs.get_train_eia().reset_index()
+    block_col = "plant_id_report_year_util_id"
+    out = ferc1_df.merge(
+        eia_df, how="inner", on=block_col, suffixes=("_ferc1", "_eia")
+    ).set_index(["record_id_ferc1", "record_id_eia"])
+    return out
+
+
+@op
+def get_y_label_df(train_pairs_df, inputs):
+    """Get the dataframe of y labels.
 
-@asset(
-    io_manager_key="pudl_sqlite_io_manager",
-    compute_kind="Python",
+    For each record pair in ``train_pairs_df``, a 0 if the pair is not
+    a match and a 1 if the pair is a match.
+    """
+    label_df = np.where(
+        train_pairs_df.merge(
+            inputs.get_train_df(),
+            how="left",
+            left_index=True,
+            right_index=True,
+            indicator=True,
+        )["_merge"]
+        == "both",
+        1,
+        0,
+    )
+    return label_df
+
+
+@op
+def get_best_matches_with_overwrites(match_df, inputs):
+    """Get dataframe with the best EIA match for each FERC record."""
+    return find_best_matches(match_df).pipe(overwrite_bad_predictions, inputs.train_df)
+
+
+@op
+def run_matching_model(features_train, features_all, y_df):
+    """Run model to match EIA to FERC records."""
+    return run_model(
+        features_train=features_train,
+        features_all=features_all,
+        y_df=y_df,
+    )
+
+
+@op
+def get_match_full_records(best_match_df, inputs):
+    """Join full dataframe onto matches to make usable and get stats."""
+    connected_df = prettyify_best_matches(
+        matches_best=best_match_df,
+        plant_parts_eia_true=inputs.get_plant_parts_eia_true(),
+        plants_ferc1=inputs.get_plants_ferc1(),
+        train_df=inputs.get_train_df(),
+    ).pipe(add_null_overrides)  # Override specified values with NA record_id_eia
+    return Resource.from_id(
+        "out_pudl__yearly_assn_eia_ferc1_plant_parts"
+    ).enforce_schema(connected_df)
+
+
+@op
+def get_pair_vectorizers():
+    """Get dictionary of vectorizers for each column in input dataframe."""
+    return pair_vectorizers
+
+
+@graph_multi_asset(
+    outs={
+        "out_pudl__yearly_assn_eia_ferc1_plant_parts": AssetOut(
+            io_manager_key="pudl_sqlite_io_manager"
+        )
+    }
 )
 def out_pudl__yearly_assn_eia_ferc1_plant_parts(
```

Comment:
Maybe some of the ops that are called in this graph asset should be combined so there are fewer tiny ops. I wasn't really sure what the best practice was here, so I made a bunch of little ops.

Reply:
There's some overhead to each `op`, so it can improve performance to combine them, but it's probably not too big of a deal as long as it's not excessive.